Add SPARQL* support to parser and generator

AlexeyMz opened the pull request:
This is a draft implementation of SPARQL* support in parser and generator from our team at metaphacts.
(It fixes #105)

I'm not very familiar with parser rules structure but I tried my best at making as little changes as possible to support triple references (`<<s p o>>`).
Also the generator currently explicitly checks if a term is a quad which won't be needed in the future (as `Quad` will implement `Term` interface).

rubensworks commented:
> N3.js can also work with triples that have quads as their subject. For example <<<a> <b> <c> <g>>> <b> <e>. I suggest adapting SPARQL to this as well.

We may want to add a config option for this. Default RDF* does not allow quads here, but when enabled via a config option we could allow this, as this may be valuable for experimentation.

Also, we may want to follow the SPARQL syntax for representing quads here, i.e. `<<GRAPH <g> { <a> <b> <c> }>> <b> <e>`

RubenVerborgh commented:
Also, it would need to be behind a flag for me, such that we can be 100% compatible with SPARQL 1.1 (i.e., reject queries tthat are not valid SPARQL 1.1, which includes SPARQL*).

AlexeyMz commented:
@simonvbrae @rubensworks I agree, this would be a valuable feature to experiment with. For now I think it would be great to have a basic syntax support without nested triples and quads; it should be easy to extend it in the near future under a flag.

@RubenVerborgh I updated the PR to use N3 v1.6.0 with RDF* support and added `allowRdfStar` flag to the parser to throw an error on `<< >>` syntax if it's not explicitly set.

simonvbrae commented:
> @simonvbrae @rubensworks I agree, this would be a valuable feature to experiment with. For now I think it would be great to have a basic syntax support without nested triples and quads; it should be easy to extend it in the near future under a flag.
> 
> @RubenVerborgh I updated the PR to use N3 v1.6.0 with RDF* support and added `allowRdfStar` flag to the parser to throw an error on `<< >>` syntax if it's not explicitly set.

I have implemented nested triples and tests. I could add this to your code and then, if @RubenVerborgh approves, this PR should be good to go.

AlexeyMz commented:
@simonvbrae It would be great if you update this PR with nested triples support 😄

RubenVerborgh commented on bin/sparql-to-json:
```diff
@@ -16,5 +16,20 @@ source.setEncoding('utf8');
-  process.stdout.write(JSON.stringify(parseTree, null, '  ') + '\n');
+  function termReplacer(key, value) {
+    if (isQuad(value)) {
+      // Workaround for N3 Quad.toJSON() missing `termType` property
```
@simonvbrae Do we need to fix this in N3?

simonvbrae replied:
Hmm I thought so too initially. I checked to see if I could find the problem in N3 but I didn't see a problem there... @AlexeyMz Why do we need this workaround?

RubenVerborgh commented:
Ah wait, we probably want to continue this in #111 instead.
